docs/proposals/secure-impersonation.md

```diff
@@ -0,0 +1,636 @@
+Restricting `[command|cmd-path]` to `/kubeconfig-bin/` allows the administrator an opt-in directory where they can add tools they trust/accept like `gcloud` and `aws-iam-authenticator` 
+while disallowing access to more dangerous binaries like `kubectl`. This mechanism is somewhat inspired by `/sbin` and `setuid` binaries.
+
+Using exec helpers is not an approach used by Cluster API.
+In addition to being insecure, it's not possible to distrubite controllers containing all of the necessary binaries to support all kubernetes platforms.
+Cluster API providers instead schedule regular updates to kubeConfig Secret data which is more tenant friendly and dependency-free.
+
+Regardless, we must sanitize the kubeConfig, so it's easy to add a restricted, PATH-based allow-list that the flux-system owner can control if they choose to manage remote clusters in this way.
+
+
+#### sourceRefs
+Controllers should impersonate the reconciler User when fetching sourceRefs.
+This allows for users to open up Source access policies via RBAC, rather than relying on a Gatekeeper or Kyverno installation to restrict default Flux behavior.
+ClusterRoles should exist for the Source controller API's `flux-source-viewer`, `flux-gitrepo-viewer`, `flux-helmrepo-viewer`, `flux-bucket-viewer`, etc.
+FluxUsers should be bound explicitly against these ClusterRoles, either with RoleBindings (for single Namespaces) or ClusterRoleBindings.
+Cross-Namespace access of Sources can be enabled per Namespace or across the whole cluster via RoleBindings/ClusterRoleBindings to a FluxUser, Flux Group, or All FluxUsers.
+Kinds of sources can be differentiated, and individual Sources can be permitted per-User/Group through `resourceNames` within Role rules.
+The current API objects that use sourceRefs are `Kustomization` and `HelmChart`.
+`Kustomization` and `HelmRelease` have the `serviceAccountName` proposed `user` fields for impersonation.
+Both of these impersonation fields should also be added to `HelmChart` because reconciling one requires accessing other Sources.
+source-controller will need impersonation behavior implemented /w additional RBAC and any supporting flags/options.
+Alternatively, `HelmChart` could be moved into the helm-controller control loops instead of source-controller.
+It's not sufficient to gate the creation of `HelmCharts` based off of a `sourceRef` permission check in `HelmRelease`;
+RBAC rules can change at any time, so the source-controller API client should have the constraints of the impersonated User.
+
+If a kubeConfig is specified for a resource that specifies a sourceRef, a separate client matching the same User or SA from the remote cluster is used in the source cluster.
+This is sensible because the User for the remote cluster must match the source cluster's reconciling Namespace.
+If no User is specified for the remote cluster, the default FluxUser is impersonated to fetch the sourceRef from the source cluster.  
+
+
```

Comment:
Is it worth talking about transitive access to secrets via sources, here? E.g., should an impersonated user need access to a secret referenced by a source. If it's a non-issue, it'd be good to at least say why it's a non-issue, to cover that off.